Logical types binary marshaling for timestamp types

samv opened the pull request:
This adds (IMHO) binary marshaling for an important logical type - time.Time.  It's important because time fields are common in event data, and this representation is efficient to produce from the internal representation of a time in go.

- [x] SpecificDatumWriter - timestamp types excl. time.Duration/time-millis, time.Time/time-millis, time.Time/date
- [x] SpecificDatumReader - timestamp types

Here's the new test with some debugging output for illustration:

```
=== RUN   TestSpecificDatumWriterLogicalTypes
2018/08/08 10:49:12 Got for long (timestamp-millis): 1970-01-02 02:17:36.789 -0800 PST
2018/08/08 10:49:12 Writing: 123456789
2018/08/08 10:49:12 Got for long (timestamp-micros): 2009-02-13 15:31:30.123456 -0800 PST
2018/08/08 10:49:12 Writing: 1234567890123456
2018/08/08 10:49:12 Got for long (time-micros): 2009-02-13 07:24:36.54321 +0000 UTC
2018/08/08 10:49:12 Writing: 26676543210
2018/08/08 10:49:12 Converted timestamp-millis value 123456789 to 1970-01-02 02:17:36.789 -0800 PST
2018/08/08 10:49:12 Converted timestamp-micros value 1234567890123456 to 2009-02-13 15:31:30.123456 -0800 PST
2018/08/08 10:49:12 Converted time-micros value 26676543210 to 1970-01-01 07:24:36.54321 +0000 UTC
```

Commit:
Add support for logical types

Logical Types are in the standard, so they should probably be supported as
first-class fields.  Add the basic schema-level support for this.

See https://avro.apache.org/docs/1.8.2/spec.html#Logical+Types

Commit:
Support Aliases in schema

Aliases are important when matching schemas to support
schema evolution etc; parse them into the schema structure
(and add the missing SchemaField struct field)

https://avro.apache.org/docs/1.8.2/spec.html#Aliases

Commit:
Support custom 'bytes' logicalTypes

Binary data is a useful base type for laying custom types
on; allow more than just the one defined in the spec.  The
spec says to treat unknown logical types as their underlying
type, so this should be standards-compliant and perhaps even
interoperable.

Commit:
Handle 'long' logical types in Specific Datum Writer

The specific data writer would blow up when writing these
types because it was expecting an underlying int64 type.
Allow the time-related logical types to marshal out from
their standard library go types.

natemurthy commented on schema.go:
```diff
@@ -292,8 +293,53 @@ func (*LongSchema) Prop(key string) (interface{}, bool) {
+			return micros
+		}
+	}
+	return dereference(v).Interface().(int64)
```
Is `dereference` a reserved keyword synonym for `*`? If so had no idea this existed.

samv replied:
It's not reserved but you tend to see a version of this in a lot of code that deals with reflection.  If the reflect.Value points to a pointer to a value, it follows that pointer and instead returns the reflect.Value pointing to the value.  Otherwise it returns the argument.

Commit:
Handle timestamp and time types in Specific Datum Reader

When reading data encoded with one of the various
'timestamp'-logical types, if the field it is being
marshaled into is not an int64 but a time.Time, then this
will panic on unmarshal.  Detect based on the presence of
the 'logicalType' field (and a bad target `reflect.Kind`) to
call into a special function which can handle mapping of
logical types to (in this case) `time.Time` values.

In some ways it would be nicer if marshaling could be
customizable eg via methods, but then you couldn't easily
add it for standard library types like `time.Time` anyway.

samv commented:
I think I'll also leave off the generic datum writer for timestamps, on the basis that it is still possible to use these logical types without it.  This is probably enough to get feedback on whether the approach/direction is agreeable to the maintainers.

samv commented on datum_reader.go:
```diff
@@ -185,6 +186,11 @@ func (reader sDatumReader) readValue(field Schema, reflectField reflect.Value, d
 		return reader.mapPrimitive(func() (interface{}, error) { return dec.ReadInt() })
 	case Long:
+		longSchema := field.(*LongSchema)
+		if longSchema.LogicalType != "" && reflectField.Kind() != reflect.Int64 {
```
I just checked and it seems that the method used to set the value requires the target field to be specifically Int64 (no other Int type is assignable).  i.e., it's not using reflect.SetInt so this should be correct.

natemurthy commented:
I think you'll want to rebase off the head of `teslamotors/avro` master so you can avoid the conflicts if you rebase and merge (or you can merge without rebase or squash and merge). Otherwise, LGTM.
